_posts/2025-11-25-variant.md

```diff
@@ -0,0 +1,114 @@
+
+## Whats the problem with JSON?
+
+Many real world datasets are messy and lack a rigid schema. As a result, it's common to store such data as a JSON column.
+
+The access pattern of this model looks like:
+
+- **read**: for every row you scan, decode the entire JSON into memory and evaluate it
+- **write**: for every record you write, encode it as its own JSON value
+
+There are many problems with this approach.
+
+_Read performance degrades quickly._ To evaluate any predicate requires scanning every row and deserializing the entire JSON payload into memory, no matter its size. Even multi-megabyte documents must be completely decoded just to inspect a single field.
+
+_Write efficiency is also poor._ Because each row stores an independently encoded JSON object, common object fields are redundantly serialized over and over. This increases storage size and adds unnecessary encoding overhead.
+
+These problems aren't new and solutions exist that improve the access patterns of a JSON column. One common solution is to extract commonly occuring object keys into a dedicated column, also known as object shredding. Yet, this requires substantial engineering effort and implementation details vary by query engine.
+
+## How Variant solves these problems
+
+Variant is a data type with an efficient binary encoding. It's designed to store JSON data in a way that is more performant for OLAP query engines.
+
+### Variant has richer data types
+
+JSON is limited to just six data types: strings, numbers, booleans, nulls, objects, and arrays. This simplicity comes at a cost, as specialized data types like timestamps and UUIDs must be encoded as strings, then inferred and parsed back at query time.
+
+Variant removes this overhead by supporting a much broader range of native types organized into three categories: primitives, objects, and arrays. Objects and arrays work exactly like their JSON counterparts, supporting arbitrary nested structures. Variant extends the primitive category to include 20 specialized types. Dates, timestamps, UUIDs, binary data, and integers and floats of various width all have their own native representations. Values are encoded in type-specific binary formats, optimized for their native type rather than a stringifed representation. This reduces both storage-overhead and query-time parsing costs.
+
+_[Figure: diagram of { id: "some uuid", timestamp: ...} encoded in both JSON and Variant logically and physically]_
+
+### Variant has efficient serialization through a 2-column design
+
+JSON columns have naive serialization which leads to unnecessary encoding overhead. For example, when storing `{"user_id": 123, "timestamp: "2025-04-21" }` thousands of times, the field names `user_id` and `timestamp` are written out in full for each row, even though they're identical across rows.
```

Comment:
this is a good example, I think we could make a nice diagram that illustrates the point nicely -- aka a bunch of JSON documents with repeated fields showing that when encoded as JSON the strings get repeated over and over

Then we can add a "here is how you would encode it as variant" showing both the single copy of field names as well as the split into two different fields 🤔